src/client/pythonEnvironments/info/workerPool.ts

```diff
@@ -0,0 +1,215 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+'use strict';
+interface IWorker {
+    stop(): void;
+    run(): void;
+}
+
+type NextFunc<T> = () => Promise<T>;
+type WorkFunc<T, R> = (item: T) => Promise<R>;
+type PostResult<T, R> = (item: T, result?: R, err?: Error) => void;
+
+interface IWorkerFactory {
+    createWorker<T, R>(next: NextFunc<T>, workFunc: WorkFunc<T, R>, postResult: PostResult<T, R>): IWorker;
+}
+
+class Worker<T, R> implements IWorker {
+    private stopProcessing: boolean = false;
+    public constructor(
+        private readonly next: NextFunc<T>,
+        private readonly workFunc: WorkFunc<T, R>,
+        private readonly postResult: PostResult<T, R>
+    ) {}
+    public stop() {
+        this.stopProcessing = true;
+    }
+
+    public async run() {
+        while (!this.stopProcessing) {
+            try {
+                const workItem = await this.next();
+                try {
+                    const result = await this.workFunc(workItem);
+                    this.postResult(workItem, result);
+                } catch (ex) {
+                    this.postResult(workItem, undefined, ex);
+                }
+            } catch (ex) {
+                // Next got rejected. Likely worker pool is shutting down.
+                // continue here and worker will exit if the worker pool is shutting down.
+                continue;
+            }
+        }
+    }
+}
+
+function createWorker<T, R>(nextItem: NextFunc<T>, workFunc: WorkFunc<T, R>, postResult: PostResult<T, R>): IWorker {
+    return new Worker<T, R>(nextItem, workFunc, postResult);
+}
+
+interface IResultHolder<R> {
+    resolve(result?: R | PromiseLike<R>): void;
+    reject(reason?: Error): void;
+}
+
+interface IWorkItem<T> {
+    item: T;
+}
+
+export enum QueuePosition {
+    Back,
+    Front
+}
+
+export interface IWorkerPool<T, R> {
+    addToQueue(item: T, queuePosition: QueuePosition): Promise<R>;
+    stop(): void;
+}
+
+export class WorkerPool<T, R> {
+    // This collection tracks the full set of workers.
+    private workers: IWorker[] = [];
+
+    // This is set by stop()
+    private stopProcessing: boolean = false;
+
+    // This is where we store
+    private queue: IWorkItem<T>[] = [];
+
+    // A collections that holds unblock callback for each worker waiting
+    // for a work item when the queue is empty
+    private waitingWorkersUnblockQueue: (() => void)[] = [];
+
+    // A map of work item and its associated promise.
+    private resultsMap: Map<IWorkItem<T>, IResultHolder<R>> = new Map();
+
+    public constructor(
+        workerFunc: WorkFunc<T, R>,
+        count: number = 2,
+        workerFactory: IWorkerFactory = { createWorker }
```

Comment:
`createWorker()` above is not needed if you inline here:

```suggestion
        workerFactory: IWorkerFactory = { createWorker: (ni, wf, pr) => new Worker<T, R>(nextItem, workFunc, postResult); }
```